cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,819 @@
+    unchanged <- liftIO $
+      checkFileModificationTimeAndHashUnchanged root file mtime hash
+    unless unchanged (somethingChanged file)
+
+-- | Within the @root@ directory, check if @file@ still does not exist.
+-- If it *does* exist, short-circuit.
+probeFileNonExistance :: FilePath -> FilePath -> ChangedM ()
+probeFileNonExistance root file = do
+    exists <- liftIO $ doesFileExist (root </> file)
+    when exists (somethingChanged file)
+
+-- | Returns @True@ if, inside the @root@ directory, @file@ has the same
+-- 'ModTime' as @mtime@.
+checkModificationTimeUnchanged :: FilePath -> FilePath
+                               -> ModTime -> IO Bool
+checkModificationTimeUnchanged root file mtime =
+  handleDoesNotExist False $ do
+    mtime' <- getModificationTime (root </> file)
+    return (mtime == mtime')
+
+-- | Returns @True@ if, inside the @root@ directory, @file@ has the
+-- same 'ModTime' and 'Hash' as @mtime and @chash@.
+checkFileModificationTimeAndHashUnchanged :: FilePath -> FilePath
+                                          -> ModTime -> Hash -> IO Bool
+checkFileModificationTimeAndHashUnchanged root file mtime chash =
+  handleDoesNotExist False $ do
+    mtime' <- getModificationTime (root </> file)
+    
+    if mtime == mtime'
+      then return True
+      else do
+        chash' <- readFileHash (root </> file)
+        return (chash == chash')
+
+-- | Read a non-cryptographic hash of a @file@.
+readFileHash :: FilePath -> IO Hash
+readFileHash file =
+    withBinaryFile file ReadMode $ \hnd ->
+      evaluate . Hashable.hash =<< BS.hGetContents hnd
+
+-- | Given a directory @dir@, return @Nothing@ if its 'ModTime'
+-- is the same as @mtime@, and the new 'ModTime' if it is not.
+checkDirectoryModificationTime :: FilePath -> ModTime -> IO (Maybe ModTime)
+checkDirectoryModificationTime dir mtime =
+  handleDoesNotExist Nothing $ do
+    mtime' <- getModificationTime dir
+    if mtime == mtime'
+      then return Nothing
+      else return (Just mtime')
+
+-- | Run an IO computation, returning @e@ if it raises a "file
+-- does not exist" error.
+handleDoesNotExist :: a -> IO a -> IO a
+handleDoesNotExist e =
+    handleJust
+      (\ioe -> if isDoesNotExistError ioe then Just ioe else Nothing)
+      (\_ -> return e)
+
+-- | Run an IO computation, returning @e@ if there is an 'error'
+-- call. ('ErrorCall')
+handleErrorCall :: a -> IO a -> IO a
+handleErrorCall e =
+    handle (\(ErrorCall _) -> return e)
+
+------------------------------------------------------------------------------
+-- Instances
+-- 
+
+instance Text FilePathGlob where
+  disp (GlobDir  glob pathglob) = disp glob Disp.<> Disp.char '/'
+                                            Disp.<> disp pathglob
+  disp (GlobFile glob)          = disp glob
+
+  parse = parse >>= \glob ->
+            (do _ <- ReadP.char '/'
+                globs <- parse
+                return (GlobDir glob globs))
+        <++ return (GlobFile glob)
```

Comment:
Indentation is a bit off, looks like `(parse >>= [...]) <++ (return [...])` instead of `parse >>= (\glob -> [...] <++ [...])`.